File: backend/app/models/curator.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class VisualEvidence(BaseModel):
    should_capture: bool = False
    reason: str | None = None
    target_time_hint: Literal["start", "middle", "end"] = "middle"


class AudioSignal(BaseModel):
    used: bool = False
    reason: str | None = None


class SourceGrounding(BaseModel):
    transcript_used: bool = False
    visual_used: bool = False
    audio_used: bool = False


class CuratorDecision(BaseModel):
    should_write: bool = True
    skip_reason: str | None = None

    concept_title: str = ""
    summary: str = ""
    key_points: list[str] = Field(default_factory=list)
    takeaway: str = ""
    confidence: float = 0.0

    visual_evidence: VisualEvidence = Field(default_factory=VisualEvidence)
    audio_signal: AudioSignal = Field(default_factory=AudioSignal)
    source_grounding: SourceGrounding = Field(default_factory=SourceGrounding)

    window_id: int = 0
    window_start: float = 0.0
    window_end: float = 0.0

File: backend-go/internal/models/curator.go
package models

// VisualEvidence describes whether and how a screenshot should be captured
// alongside a curator decision.
type VisualEvidence struct {
	ShouldCapture bool `json:"should_capture"`
	Reason        string `json:"reason"`
	// TargetTimeHint controls which frame within the window is preferred.
	// Valid values: "start", "middle", "end". Defaults to "middle".
	TargetTimeHint string `json:"target_time_hint"`
}

// AudioSignal describes whether audio context was meaningful for the
// curator decision.
type AudioSignal struct {
	Used   bool   `json:"used"`
	Reason string `json:"reason"`
}

// SourceGrounding records which input modalities were used when reaching
// the curator decision.
type SourceGrounding struct {
	TranscriptUsed bool `json:"transcript_used"`
	VisualUsed     bool `json:"visual_used"`
	AudioUsed      bool `json:"audio_used"`
}

// CuratorDecision is the structured output produced by the AI curator for
// a single EventWindow. It determines whether the window's content should
// be written to Notion and carries the distilled knowledge payload.
type CuratorDecision struct {
	ShouldWrite bool   `json:"should_write"`
	SkipReason  string `json:"skip_reason"`
	ConceptTitle string `json:"concept_title"`
	Summary     string `json:"summary"`
	KeyPoints   []string `json:"key_points"`
	Takeaway    string   `json:"takeaway"`
	Confidence  float64  `json:"confidence"`

	VisualEvidence  VisualEvidence  `json:"visual_evidence"`
	AudioSignal     AudioSignal     `json:"audio_signal"`
	SourceGrounding SourceGrounding `json:"source_grounding"`

	// Window provenance — echoed from the originating EventWindow.
	WindowID    int     `json:"window_id"`
	WindowStart float64 `json:"window_start"`
	WindowEnd   float64 `json:"window_end"`
}

// NewCuratorDecision returns a CuratorDecision with its zero-safe defaults
// applied: ShouldWrite is true and VisualEvidence.TargetTimeHint is "middle".
func NewCuratorDecision() CuratorDecision {
	return CuratorDecision{
		ShouldWrite: true,
		VisualEvidence: VisualEvidence{
			TargetTimeHint: "middle",
		},
	}
}
